### DISCUSSION TITLE
Notebooks

### DISCUSSION THREAD

**Discussion by Milo123459** [minbrowser/min#1451] [Ideas]
Hello! Me and palmer have been talking about Min Notebooks, a way to store information, and well, notes.

The main idea:
A space dedicated for users to put down information / notes on a website / multiple ones. For example, min://notebooks is a GUI allowing you to create / view / search for notebooks. You can take screenshots of pages and paste them in to the notebook. It's fully markdown supported.

Some of my other ideas:
Git hosted. Login with a github account and have it synced on a private repo. You could also make a VSC extension for this, but, these are prototyped ideas :)

Let me know of some thoughts, and we'll see what we can do 🗡️

**Comment by PalmerAL** [comment]
I think this is a cool idea :)

Something we were talking about is having this integrated with tasks / history, so your notebook would be a combination of any pages you visit and the notes you take on them.

Syncing would be good, but probably not through git - it's fairly complicated to set up, you can't easily delete content (if you accidentally put private content in your notebook), and it would make it harder to have shareable notebooks if we ever wanted to do that. Probably better to build a custom sync service (maybe charge a couple $/month per user to run it?)

**Reply by IdleEndeavor** [reply]
Yes I agree with you. They might have to charge for sync but keep the private on device use free. This would greatly increase the productivity part of Min Browser.

**Reply by Milo123459** [reply]
Git Sync is something to get your head around without a doubt, but, it's very possible. Some of the other challenges are we'd have to make it work without a local git environment, which, is a bit annoying. As of right now, I'm planning the rewrite for min (can be found [here](https://github.com/minbrowser/min/discussions/1447)) so this is low priority. Please let me know of suggestions tho!

**Reply by Milo123459** [reply]
On the custom sync, that's probably a good idea. Personally, I think keeping it free is the best idea, but that's subject to change.
This is how I picture it'd work:
- Sign up on the minbrowser website or something, accounts would be required for security
- Sign in in the actual min browser application
- Makes get requests to the API when opening up a notebook but it is also stored locally, if conflicts exists it'll warn you.
- Edit something, press save and it makes the post request.

**Reply by blackgwe** [reply]
I just add my thoughts here: #1518 (contains screenshots for better understanding)

Main idea is to integrate / rebuild something like https://hypothes.is/ -> [Here](https://via.hypothes.is/https://www.inboundlogistics.com/cms/article/shipment-monitoring-technology-picking-up-the-signals/#annotations:1dl7inseEeu2fheC53Za8A) you can see tags on individual highlights / annotations.

see also the W3C Web Annotation: https://w3c.github.io/web-annotation/

**Comment by PalmerAL** [comment]
Related article: https://sifted.eu/articles/beam/

**Reply by blackgwe** [reply]
related source: https://en.wikipedia.org/wiki/Web_annotation

**Reply by blackgwe** [reply]
maybe it could be easy to integrate it with wallabag (nice API)
- https://www.youtube.com/watch?v=dti8IiczRSI
- https://github.com/wallabag/wallabag
- https://doc.wallabag.org/en/

@PalmerAL what do you think about it?

**Reply by PalmerAL** [reply]
Hmm, not sure what I think, but I wrote down some thoughts in a separate comment: https://github.com/minbrowser/min/discussions/1451#discussioncomment-643269

**Comment by PalmerAL** [comment]
One other possibility (brought up in #1518, and suggested by @blackgwe) is to have an integration with a read-it-later service, like Pocket/Pinboard/Wallabag/others.

We could integrate with these services by creating a userscript (so you could install the userscript > right-click "save to <service>"). I think that would work pretty well. In some cases, it may require us to fix web compat issues (such as Hypothes.is, which doesn't work because of issues with our popup window support), but those are things we should fix anyway.

We could also build some kind of native integration into the browser. The way I'm imagining that is that when you enable the read-it-later service, your existing local bookmarks in Min get uploaded, and then the bookmarks manager just becomes a view into whatever's saved in the read-it-later service. If we added native annotation features, they would similarly upload to the service.

This might have a lower upfront cost, because it lets us re-use someone else's storage/sync infrastructure instead of building our own. However, it introduces a new set of issues:

* The market for read-it-later services is pretty fragmented - there's Pocket, Pinboard, Wallabag, Hypothesis, etc. By picking one of these, we're potentially forcing people to switch services against their will or picking a service that won't survive long-term, but if we try to support multiple services, our implementation complexity increases a lot.
  * We ran into a similar issue with password managers, and there we've tried to support connecting to multiple third-party services through a shared UI in Min. It's sort of worked, but it's ended up being a lot of work to maintain, the third-party integrations are sort of buggy, and adding features is hard. I'm not sure I want to repeat that again.
* By having a native integration, we'd be seen as endorsing a particular service, even though we don't really have any control over their features/design/etc.
  * What happens if the third-party service has a security issue?
  * What happens if the third-party service decides to start charging for API access? (Maybe not as much of an issue with Wallabag, since you can self-host)
* We're limited to whatever feature set the third-party supports. Even this original discussion is about notebooks, not annotations - unless Wallabag or some other service had notebook support, we couldn't sync it to their service.

On the other hand, it appears that we don't have enough users to justify building and running our own sync service.

As a final note: do we actually need sync at all? Sync is typically useful for going from mobile <-> desktop, but we don't have a mobile version anyway, so sync would just be limited to people with multiple desktop computers. How many people is that? We also already have client-side full-text history search, so we don't need a server integration to do that.

(I suppose if we integrated with a service like Wallabag, it would have mobile clients, so it might be somewhat more useful than our own sync service).

**Reply by blackgwe** [reply]
**Help wanted** I'd like to start in summer to integrate an external service using the proposed userscript mechanism.(probably wallabag – open source/ self hosted or in the cloud, mobile clients for reading later/offline and annotation) 

If somebody else wants to contribute (code, documentation, ideas, improvements) – you are welcome!
If another similar solution is proposed, I'd like to discuss / help and maybe merge the (sub-) projects together. 

@PalmerAL THANKS a lot for your ideas and the efforts to steadily improve the `min` browser and keep it as min as possible!

**Reply by PalmerAL** [reply]
@blackgwe I wrote a userscript for Hypothes.is: https://github.com/PalmerAL/min-hypothesis-userscript. You'll need Min 1.20 or later (released as a beta yesterday).

I think it should be possible to use an almost identical script for Wallabag, since they supposedly have a bookmarklet as well. I didn't try it, since you have either pay or install it yourself to get to the bookmarklet installation step.

**Reply by yurikhan** [reply]
> sync would just be limited to people with multiple desktop computers. How many people is that?

Pretty much everyone who has an office job?

**Comment by raphael10-collab** [comment]
@PalmerAL It seems that `Apache Annotation` project is trying to unify under the same "umbrella"-project all the different annotation approaches: https://annotator.apache.org/ https://github.com/apache/incubator-annotator
